design/0005-circuit-breaking.md

```diff
@@ -0,0 +1,231 @@
+specific service.
+
+[Example](https://drive.google.com/open?id=1jSMaogdvzxjv0v2dBoL7OYGHmsCusXn-)
+
+---
+
+Network Parition - There can be lot of scenarios in network partition but to
+provide a glimpse how circuit breaking would behave in case of network issue -
+[Example](https://drive.google.com/open?id=154I2ZGlGc-rQJ0YVeek17dJy7raAvd-G)
+
+---
+
+Protocol Specific features -
+
+1. For HTTP/gRPC based communication, users should be able to configure circuit
+   breaking either per endpoint or whole service.
+   [Example](https://drive.google.com/open?id=1JK65u9Mnmkmy_X8fczeh6RO6O-vNiTtY)
+   User can configure ejection based on many different metrics. These metrics
+   would be calculated over a given time range.
+   - Success Rate
+   - Percentile Latencies
+   - Max Requests
+   - Consecutive Failures
+2. For TCP based communication -
+   [Example](https://drive.google.com/open?id=1yAAurygsZQ-r58TEA0QiFf9Sz15jqRJV)
+
+---
+
+[0] The term 'traffic' means any kind of communication happening between two
+services. It can be HTTP requests or gRPC messages over HTTP or L4 TCP
+connections. But the circuit breaking functionality itself won't be protocol
+agnostic. Its implementation would be protocol specific and it's upto community
+which protocol they want to add support for and how circuit breaking will make
+sense in that protocol.
+
+## Design proposal
+
+[design-proposal]: #design-proposal
+
+Right now if I just consider ServiceProfile spec, IMO it seems to represent more
+of profile-for-http-service rather than profile-for-generic-service. May be it
+makes sense given that it supports only HTTP and grpc over HTTP as of now. Given
+current architecture, I feel it's best to provide circuit breaking functionality
+through service profiles. That means it would be limited to above two kind of
+services as of now. Circuit breaking logic would be applied in the data plane
+based on the metrics recorded inside the proxy.
+
+Since circuit breaking needs to be handled per protocol, we will leave TCP for
+now and can revisit the same in future.
+
+In future, for any other protocol support for circuit breaking, one will first
+need to incorporate the protocol into ServiceProfile itself and then implement
+the circuit breaking functionality.
+
+For HTTP/gRPC, we already have golden metrics available in the proxy. Along with
+that, we can add additional metrics and use those for circuit breaking logic. In
+case no instance is available to serve request due to circuit breaking, then we
+will return 503 for HTTP with some custom header and gRPC message for gRPC.
+
+ServiceProfile interface would have following additions for the HTTP and gRPC -
+
+### Circuit Breaker
+
+```yaml
+routes:
+  - name: "/foo"
+    circuitBreaker:
+      # Circuit Breaker Kind
+      # (Required)
+      kind: ...
+
+      # Max percentage of instances that can be ejected.
+      # Default - 100%
+      maxEjectionPercentage: 10
```

Comment:
We are going to eject instance completely from the upstream instance rather than having decreasing weights for unhealthy instances. So, in that approach, having percentages is way to go. 

But as mentioned [here](https://github.com/linkerd/rfc/pull/30/files#r461794191), I haven't been able to provide convincing explanation for the same, so I'll move it to out of scope for now.